### PR TITLE
Synchronize operator

### PR DESCRIPTION
**Pull request by abarret**
@bindi-nagda take a look at this branch. It introduces a couple of options to synchronize different parts of the Krylov operator. I can get the Krylov solver to converge to a residual of `1e-15` in almost all cases with these options. Running the four-roll-mill target gives a non-converging solver sometimes, I'm unsure why. Check and make sure that you can too, and then see what order of accuracy you get with the solver. I'm also unsure if these are exactly the correct steps to take, I'm still exploring what exactly these changes do.

Note this requires the most recent IBAMR main commit, so you will probably have to pull in the IBAMR build.

### CODE REVIEW

File: VCTwoFluidStaggeredStokesOperator.cpp [hunk 1]
```diff
@@ -325,6 +391,24 @@ VCTwoFluidStaggeredStokesOperator::apply(SAMRAIVectorReal<NDIM, double>& x, SAMR
     StaggeredStokesPhysicalBoundaryHelper::resetBcCoefObjects(d_un_bc_coefs, d_P_bc_coef);
     d_hier_bdry_fill->resetTransactionComponents(d_transaction_comps);
 
+#define USE_DIV
+#define USE_SYNCHED_INTERP
+#define POST_SYNCH
```
Comment (abarret) [inline VCTwoFluidStaggeredStokesOperator.cpp:396]: In particular, these three lines control which changes are used:

- `USE_DIV` synchronizes the side centered values before computing the divergence of the velocities.
- `USE_SYNCHED_INTERP` uses a synchronized node and side interpolated version of the volume fraction
- `POST_SYNCH` performs an extra synchronization step at the end of the `apply()` function.

You can switch back to the old behavior by commenting out one of these options and rebuilding.

**Comment by bindi-nagda** [comment]
Okay great, I will test this.

**Comment by abarret** [comment]
> Running target fluid_multigrid without preconditioning, I also get convergence to 1e-15 with rectangular and L-shaped refinement. With preconditioning turned on, the convergence is very slow. Do you also see this?

No. I see convergence with the preconditioner in ~50 iterations in L-shaped domains. Can you send me the input file you are using?

**Comment by bindi-nagda** [comment]
The solver converges in about 39-78 iterations for most test problems with L-shaped refined regions. There is stalling observed at a relative residual of 1e-04 for `input2d.var_theta.trans_vel`. 

So far, I also get convergence for the four roll mill problem. The solver takes ~30 iterations to achieve a relative residual of 1e-15 in each time step.

**Comment by bindi-nagda** [comment]
> In particular, these three lines control which changes are used:
> 
>     * `USE_DIV` synchronizes the side centered values before computing the divergence of the velocities.
> 
>     * `USE_SYNCHED_INTERP` uses a synchronized node and side interpolated version of the volume fraction
> 
>     * `POST_SYNCH` performs an extra synchronization step at the end of the `apply()` function.
> 
> 
> You can switch back to the old behavior by commenting out one of these options and rebuilding.

I suppose next we want to toggle between these options to identify which ones are helping with the convergence?

**Comment by abarret** [comment]
> I suppose next we want to toggle between these options to identify which ones are helping with the convergence?

I expect `USE_SYNCHED_INTERP` is not necessary, but the others should be. It would be good to try all combinations on an L-shaped domain + variable theta.

File: VCTwoFluidStaggeredStokesOperator.cpp [hunk 2]
```diff
@@ -60,12 +60,10 @@ static const int SIDEG = 1;
...
-static const std::string CC_DATA_REFINE_TYPE =
-    "CONSERVATIVE_LINEAR_REFINE"; // how to fill in fine cells from coarse cells, how to fill ghost cells on refine
-                                  // patch
-static const std::string SC_DATA_REFINE_TYPE =
-    "CONSERVATIVE_LINEAR_REFINE";              // how to fill in fine cells from coarse cells, how to fill ghost
-                                               // cells on refine patch
+static const std::string CC_DATA_REFINE_TYPE = "NONE"; // how to fill in fine cells from coarse cells, how to fill ghost
+                                                       // cells on refine patch
+static const std::string SC_DATA_REFINE_TYPE = "NONE"; // how to fill in fine cells from coarse cells, how to fill ghost
+                                                       // cells on refine patch
```
Comment (bindi-nagda) [inline VCTwoFluidStaggeredStokesOperator.cpp:66]: Why do we set these to none? I thought we wanted to fill in ghost cells near the coarse-fine interfaces using conservative_linear_refine.

Comment (abarret) [reply VCTwoFluidStaggeredStokesOperator.cpp:66]: The answer is in the next line. The refine operators fill in ghost cells from coarse data. We want to use coarse-fine interpolation to fill ghost cells which uses data from coarse and fine cells.

 I think that changing these from `"NONE"` to `"CONSERVATIVE_LINEAR_REFINE"` does nothing as long as `USE_CF_INTERPOLATION = true`. You can check that.

However, I don't think `USE_CF_INTERPOLATION = true` will fill in corner ghost cells, so if those are required, then using `"CONSERVATIVE_LINEAR_REFINE"` is needed.

**Comment by abarret** [comment]
I'm going to merge this patch as is so I can add more improvements. I've highlighted my concerns in #46.

**Pull request merged**